### ISSUE TITLE
Add system permission to alter an order that is tied to an IST

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#495]
_Imported from trac ticket 495.
Created by **kbower**
Opened in trac: Dec 13, 2010, 4:28:26 PM
Last modified in trac: Nov 13, 2013, 9:12:36 AM_
***
see #349

Subject: Re: Locking orders that are on ISTs

Wayne,
      Few questions:
      * Should an order be "locked" as soon as an IST_LN$SO_LN record exists or do we wait until the IST is trucked?
      * Should just the IST'ed lines be locked or should the whole order be locked?
      * Are there any fields that are alright to be edited for IST'ed orders?  Can lines be added/voided?  Ship address? Delivery date? Zone? Etc.
Thanks,
Kent
----
Kent,
Good morning, only someone who has the 'SOCTRK' security should have
the ability to update an order that is trucked.
Answers to your questions:
1.        An order line should be locked once an IST_LN$SO_LN record exist
2.        Just the lines should be locked
3.        Orders on an IST can have all fields edited except the following without the appropriate security:
        a. Someone cannot add a line
        b. Someone cannot void the order
        c. Someone cannot change the quantity on the line attached to the IST
        d. Someone cannot change the location on the line attached to the IST
        e. Someone cannot change the Pickup Delivery Flag
        f. Someone cannot change the original zone on the sale to something different
        g. Someone cannot change the delivery date to a date prior to the IST Transfer Date
Thanks,
Wayne Liburd
----
Thank you Wayne,
     Can someone with the proper security perform all a. - g.?  (even g.?)
Also,
Do you need this security to void an IST'ed line?
Do you need this security to void an non IST'ed line?
Kent
----
Kent,
In short, yes to all three your questions.
Thanks,
Wayne Liburd

**Comment by kbower** [comment]
_**rneville** trac comment on Dec 27, 2010, 5:07:34 PM:_

(In [[1000]](../commit/4b1deb5c8c50abdf015098377101fd5bb650fcd5))
closes #349 and addresses #494 and addresses #495 in that I've added a new component to sit between the order header and the customer section called the OrderStatusBar...this will contain icons, etc. for things like Trucked or on an IST and other things could be put there. The other piece of the puzzle is that a generic enable/disable flag can be rippled through the OrderHeaderEdit/CustomerEdit and OrderLineEdit components.

Also, REVISION 1000 BABY!!!! Yahooo!!!

**Comment by kbower** [comment]
_**rneville** trac comment on Dec 29, 2010, 11:25:06 AM:_

Kent's going to implement a field-level authorization framework

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 5, 2011, 3:24:29 PM:_

(In [[1017]](../commit/b748e1c2c75424567058b8f527689024c5565a2c))
 * Implement `restricted_attributes` framework (references #494 & #495)
 * Only allow "adminstation" (workstation setup) permission to setup the Site & Cashdrawer, etc. (fixes [#528](528))
 * fix issue where fieldauth for saleprice thought it was correct to throw out the fieldauth because it technically had changed at the OrderDetail level.  It had changed because its value had been set from the OrderLine.  Therefore changeset included 'saleprice' correctly (and needs to so that autoupdates know this needs value had changed and needs to be returned updated for the client).  To fix this, fieldauth is inspecting the current value versus the passed_obj (or, in this case passed_line_obj) and so may be different than changeset for orderdetails.
 * add `for_client_write` parameter for query controller (actually, rename `usegeteagers` parameter for this use case), which means that the return from the query may be subject to client write and therefore must look like after a get (needs restricted_attributes, get joinedloads, etc.)
 * Misc
   * python
     * added `check_for_rule=False` parameter for `@rule_property()`, which allows a `@rule_property` to function as a property and for autoupdates, but without checking the database for a rule override (for cases we know there is no rule or else if there is such a function that should be more framework and probably shouldn't be subject to override)
     * added `@no_serialize` decorator for controllers
     * move the invocation of `self._check_workflows()` earlier (before set_defaults)
   * flex
     * customer query from order needs `for_client_write=true`, but from customer edit screen does not, as a subsequent get is issued.
     * refactor username password storage so it is removed from SessionAuth and set as part of the SessionAuth.user object (now stored encrypted).
       * also clear user at any failed login.  (It is probable this will fix the known bug that sometimes you could login with any garbage password.  I assume the SessionAuth.user somehow wasn't being cleared)
     * RestrictableComponent
       * Combine fieldauth and RestrictableComponent implementations (one include file)
       * add `AttrRestrictedLinkButton` and `AttrRestrictedButton`
     * For security purposes, effectively log the user out if the following failed to load at browser startup
       * dependencyCache
       * fieldAuthCache
       * systemParameters
         * systemParameter that doesn't exist
     * according to my rough benchmarks, describeType() is over 10,000 times more expensive than getQualifiedClassName(), so only invoke describeType() when really needed
       * optimize identical() function for performance
       * optimize json serialization so not invoked with plain Object type. (such as `restricted_attributes`)
     * fix empty object serialization so it renders `{}` instead of `{null}`
     * make httpservice requests more universally consistent (use params Object instead of individual Request parameters set declaratively with EventMaps from Event).  This also fixes a problem with dynaforms not really passing 'limit' etc, see HTTPService change with [r1010](../commit/952e416ddc2f2a036c22a07e7fa8103f03281326)

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 14, 2011, 10:56:28 AM:_

(In [[1052]](../commit/f5e80d046f926b44803d181378d356a143ec1094))
 * fix recursive _set_original problem caused by OrderDetail._record_change referencing self.order
   * added ability to disable_changeset 
 * fix some tables.py 
   * Date() => Date 
   * sequence datatypes that should be Unicode
   * primary key Integers need to be autoincrement=False for GERS columns
 * add ISTs (Transfers) (references #495) for future work with #495

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 15, 2011, 7:45:46 PM:_

(In [[1063]](../commit/aeec585702b70faa1e6de558a2eee2ac6b7b380d))
 * restrict changes for transferring orders (ISTs) closes #495
 * remove lead time logic from atp calculations as the lead time is already reflected by the arrivaldate (see #332)
 * minor mockdata improvements

**Comment by kbower** [comment]
_**kbower** trac status change on Jan 15, 2011, 7:45:46 PM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 15, 2011, 9:20:36 PM:_

Addendum:
Kent,

No, they should not be able to change the from site.

Thanks,

Wayne

**Comment by kbower** [comment]
_**kbower** trac comment on Apr 7, 2011, 12:12:59 PM:_

_Note, I asked JR concerning whether orderdetailtransferdetails_table needed to be many to many, but he didn't answer.  We've implemented as many to many.  See its tables.py comment.  Here is the unanswered email:_

[Edit]: See [ticket:1676#comment:12]
-------- Original Message --------
Subject:         IST_LN$SO_LN
Date:         Thu, 13 Jan 2011 07:46:05 -0500
From:         Kent Bower <kb@retailarchitects.com>
To:         Jesus Capo <jesus@eldoradofurniture.com>
CC:         Rob Neville <rn@retailarchitects.com>


JR,
    No problem if you are too busy with your store opening to answer, just at your convenience:

The structure of IST_LN$SO_LN seems to suggest to Rob and me that there is a many to many relationship possible:
```
  Name                       Null?    Type
 -------------------------- -------- -------------------
 DEL_DOC_NUM                NOT NULL VARCHAR2(14)
 DEL_DOC_LN#                NOT NULL NUMBER(4)
 IST_WR_DT                  NOT NULL DATE
 IST_STORE_CD               NOT NULL VARCHAR2(2)
 IST_SEQ_NUM                NOT NULL VARCHAR2(4)
 IST_LN#                    NOT NULL NUMBER(4)
 DIST                       NOT NULL VARCHAR2(1)
 SO_LN_SEQ                           NUMBER(4)
 SE_PART_SEQ                         NUMBER(4)
```
Specifically, the presence of field SO_LN_SEQ seems to imply that DEL_DOC_NUM and DEL_DOC_LN# alone wasn't enough. (Furthermore, one of the indexes for the table is DEL_DOC_NUM, DEL_DOC_LN#, SO_LN_SEQ, SE_PART_SEQ)

Say you have a line of qty 4 chairs for an SO_LN.  We are wondering if GERS is setup to allow up to 4 different IST lines to transfer those 4 chairs, for example:
(SO_LN_QTY: 4, each IST_LN_QTY: 1)
```
SO   SO_LN   SO_LN_SEQ  IST    IST_LN 
==== =====   =========  ====   ======
ABCD 1       1          ISTA   10
ABCD 1       2          ISTA   11
ABCD 1       3          ISTB   5
ABCD 1       4          ISTB   6
```When attaching IST lines to SO lines, _unless GERS breaks an SO line qty at the time of writing the ISTs, you would need such an arrangement, wouldn't you?_

Or what about the other way around?  Could an IST line with QTY 10 supply  up to 10 different SO_LNs (each only had qty 1)?

Does GERS allow either or both of these or will it always work one to one, even if that means breaking the SO_LN qty from qty 4 into separate lines for the SO_LN when attaching them to IST lines?


Thanks and good luck with your store grand opening!

Kent
